File: tests/rustdoc-js/assoc-type-backtrack.js
// exact-check

const EXPECTED = [
    {
        'query': 'mytrait, mytrait2 -> T',
        'correction': null,
        'others': [
            { 'path': 'assoc_type_backtrack::MyTrait', 'name': 'fold' },
            { 'path': 'assoc_type_backtrack::Cloned', 'name': 'fold' },
        ],
    },
    {
        'query': 'mytrait<U>, mytrait2 -> T',
        'correction': null,
        'others': [
            { 'path': 'assoc_type_backtrack::MyTrait', 'name': 'fold' },
            { 'path': 'assoc_type_backtrack::Cloned', 'name': 'fold' },
        ],
    },
    {
        'query': 'mytrait<Item=U>, mytrait2 -> T',
        'correction': null,
        'others': [
            { 'path': 'assoc_type_backtrack::MyTrait', 'name': 'fold' },
            { 'path': 'assoc_type_backtrack::Cloned', 'name': 'fold' },
        ],
    },
    {
        'query': 'mytrait<T>, mytrait2 -> T',
        'correction': null,
        'others': [],
    },
    {
        'query': 'mytrait<Item=T>, mytrait2 -> T',
        'correction': null,
        'others': [],
    },
    {
        'query': 'mytrait<T> -> Option<T>',
        'correction': null,
        'others': [
            { 'path': 'assoc_type_backtrack::MyTrait', 'name': 'next' },
        ],
    },
    {
        'query': 'mytrait<Item=T> -> Option<T>',
        'correction': null,
        'others': [
            { 'path': 'assoc_type_backtrack::MyTrait', 'name': 'next' },
        ],
    },
    {
        'query': 'mytrait<U> -> Option<T>',
        'correction': null,
        'others': [
            { 'path': 'assoc_type_backtrack::Cloned', 'name': 'next' },
        ],
    },
    {
        'query': 'mytrait<Item=U> -> Option<T>',
        'correction': null,
        'others': [
            { 'path': 'assoc_type_backtrack::Cloned', 'name': 'next' },
        ],
    },
];

File: tests/rustdoc-js/assoc-type-backtrack.rs
pub trait MyTrait2<X> {
    type Output;
}

pub trait MyTrait {
    type Item;
    fn next(&mut self) -> Option<Self::Item>;
    fn fold<B, F>(self, init: B, f: F) -> B where
        Self: Sized,
        F: MyTrait2<(B, Self::Item), Output=B>;
}

pub struct Cloned<I>(I);

impl<'a, T, I> MyTrait for Cloned<I> where
    T: 'a + Clone,
    I: MyTrait<Item = &'a T>
{
    type Item = T;
    fn next(&mut self) -> Option<Self::Item> { loop {} }
    fn fold<B, F>(self, init: B, f: F) -> B where
        Self: Sized,
        F: MyTrait2<(B, Self::Item), Output=B>
    {
        loop {}
    }
}
